File: 3066-minimum_operations_to_exceed_threshold_value_ii/ruby/main.rb
require 'algorithms'

# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer}
def min_operations(nums, k)
  heap = Containers::MinHeap.new
  nums.each { |num| heap.push(num) }
  operations_count = 0
  while heap.size > 1 && heap.min < k
    first_smallest = heap.pop
    second_smallest = heap.pop
    new_element = first_smallest * 2 + second_smallest
    heap.push(new_element)
    operations_count += 1
  end
  operations_count
end

if __FILE__ == $0
  nums = [2, 11, 10, 1, 3]
  k = 10
  result = min_operations(nums, k)
  puts "Result: #{result}"
  if result == 2
    puts "Passed"
  else
    puts "Failed"
  end
end

File: 3066-minimum_operations_to_exceed_threshold_value_ii/go/main.go
package main

import (
	"container/heap"
	"fmt"
)

type IntHeap []int

func (h IntHeap) Len() int           { return len(h) }
func (h IntHeap) Less(i, j int) bool { return h[i] < h[j] }
func (h IntHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *IntHeap) Push(x any) {
	*h = append(*h, x.(int))
}

func (h *IntHeap) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

func minOperations(nums []int, k int) int {
	numsHeap := IntHeap(nums)
	queue := &numsHeap
	heap.Init(queue)
	var nbOp int
	for queue.Len() > 1 && (*queue)[0] < k {
		x := heap.Pop(queue).(int)
		y := heap.Pop(queue).(int)
		heap.Push(queue, min(x, y)*2+max(x, y))
		nbOp++
	}
	return nbOp
}

func main() {
	nums := []int{2, 11, 10, 1, 3}
	k := 10
	result := minOperations(nums, k)
	fmt.Printf("Result: %#v\n", result)
	if result == 2 {
		fmt.Println("Passed")
	} else {
		fmt.Println("Failed")
	}
}
